fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunScheduler.java

```diff
@@ -0,0 +1,201 @@
+    // so the map value is priority queue need to be sorted by priority from large to small
+    private final Map<Long, Queue<TaskRun>> pendingTaskRunMap = Maps.newConcurrentMap();
+    // pending TaskRun Queue, compared by priority and created time
+    private final PriorityBlockingQueue<TaskRun> pendingTaskRunQueue = new PriorityBlockingQueue<>();
+
+    // taskId -> running TaskRun, for each Task only support 1 running taskRun currently,
+    // so the map value is not queue
+    private final Map<Long, TaskRun> runningTaskRunMap = Maps.newConcurrentMap();
+
+    /**
+     * Get the count of pending task run
+     */
+    public long getPendingQueueCount() {
+        return pendingTaskRunQueue.size();
+    }
+
+    /**
+     * Get the count of running task run
+     * @param taskId: task id
+     */
+    public TaskRun getRunningTaskRun(long taskId) {
+        return runningTaskRunMap.get(taskId);
+    }
+
+    /**
+     * @param taskId: task id
+     * @return: pending task run queue
+     */
+    public Queue<TaskRun> getPendingTaskRunsByTaskId(long taskId) {
+        return pendingTaskRunMap.get(taskId);
+    }
+
+    /**
+     * Get the pending task run queue
+     */
+    public Queue<TaskRun> getPendingTaskRuns() {
+        return pendingTaskRunQueue;
+    }
+
+    /**
+     * Add a task run to pending queue
+     * @param taskRun: task run
+     * @return: true if add success, false if add failed
+     */
+    public boolean addPendingTaskRun(TaskRun taskRun) {
+        if (taskRun == null) {
+            return false;
+        }
+        if (!pendingTaskRunQueue.offer(taskRun)) {
+            return false;
+        }
+        if (!pendingTaskRunMap.computeIfAbsent(taskRun.getTaskId(), ignored -> new PriorityBlockingQueue<>()).add(taskRun)) {
+            pendingTaskRunQueue.remove(taskRun);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Remove a task run from pending queue
+     * @param taskRun: task run
+     * @return true if remove success, false if remove failed
+     */
+    public boolean removePendingTaskRunFromQueue(TaskRun taskRun) {
+        if (taskRun == null) {
+            return false;
+        }
+        return pendingTaskRunQueue.remove(taskRun);
+    }
+
+    /**
+     * Return the current running task run map.
+     */
+    public Map<Long, TaskRun> getRunningTaskRunMap() {
+        return runningTaskRunMap;
+    }
+
+    /**
+     * Return the current pendign task run map.
+     */
+    public Map<Long, Queue<TaskRun>> getPendingTaskRunMap() {
+        return pendingTaskRunMap;
+    }
+
+    /**
+     * schedule the pending TaskRun that can be run into running TaskRun map
+     * @param action: the action to run the task run before task runs queue is full
+     */
+    public void scheduledPendingTaskRun(Consumer<TaskRun> action) {
+        int currentRunning = runningTaskRunMap.size();
```

Comment:
it seems that scheduledPendingTaskRun will be scheduled concurrently?
Is there any problem when it run concurrently?
Are there two task run for one task?

Reply:
no. it will acquire lock and should only run in the scheduler thread.

```
    public void start() {
        if (isStart.compareAndSet(false, true)) {
            clearUnfinishedTaskRun();
            registerPeriodicalTask();
            dispatchScheduler.scheduleAtFixedRate(() -> {
                if (!taskRunManager.tryTaskRunLock()) {
                    LOG.warn("TaskRun scheduler cannot acquire the lock");
                    return;
                }
                try {
                    taskRunManager.checkRunningTaskRun();
                    taskRunManager.scheduledPendingTaskRun();
                } catch (Exception ex) {
                    LOG.warn("failed to dispatch task.", ex);
                } finally {
                    taskRunManager.taskRunUnlock();
                }
            }, 0, 1, TimeUnit.SECONDS);
        }
    }
```